### PR TITLE
msglist: Share recipient headers across messages

### PR DESCRIPTION
**Pull request by gnprice**
Fixes: #174 

Screenshot with recipients changing:
Screenshot from 2023-09-07 12-47-47

Screenshot with the date changing:
Screenshot from 2023-09-07 12-48-03

### CODE REVIEW

**Review by chrisbobbe — COMMENTED** [review]
Great! See one comment below, on the tests.

File: test/model/message_list_test.dart [hunk 1]
```diff
@@ -461,6 +461,182 @@ void main() async {
     check(model).messages.length.equals(31);
     check(model.contents[0]).equalsNode(correctContent);
   });
+
+  test('recipient headers are maintained consistently', () async {
+    // This tests the code that maintains the invariant that recipient headers
+    // are present just where [canShareRecipientHeader] requires them.
+    // In [checkInvariants] we check the current state against that invariant,
+    // so here we just need to exercise that code through all the relevant cases.
+    // Each [checkNotifiedOnce] call ensures there's been a [checkInvariants] call.
```
[the remark this answers is not in the document]
Comment (gnprice) [reply test/model/message_list_test.dart:470]: Right. But the only way for `notifiedCount` to have been set greater than zero is in this listener:
```dart
    model = MessageListView.init(store: store, narrow: narrow)
      ..addListener(() {
        checkInvariants(model);
        notifiedCount++;
      });
```
and that calls `checkInvariants`.

So if `checkNotifiedOnce` succeeds (or `checkNotified` with any positive argument), then there must have been a call to that listener and it must have successfully called `checkInvariants`.

**Comment by gnprice** [comment]
Thanks for the review! Expanded that comment to answer your question.

File: test/model/message_list_test.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply test/model/message_list_test.dart]: Oh! OK, yeah, that makes sense.

**Comment by chrisbobbe** [comment]
Great, thanks! Merging.

**Pull request merged**